wallet-new/src/Cardano/Wallet/Kernel/Transactions.hs

```diff
@@ -223,6 +234,152 @@ newUnsignedTransaction ActiveWallet{..} options accountId payees = runExceptT $
+        Right (db, unsignedTx, _availableUtxo) -> runExceptT $ do
+            -- Now we have to generate the change addresses needed,
+            -- because 'newUnsignedTransaction' function cannot do it by itself.
+            changeAddresses <- withExceptT NewTransactionErrorCreateAddressFailed $
+                genChangeOuts (utxChange unsignedTx)
+                              srcAccId
+                              spendingPassword
+                              (walletPassive activeWallet)
+            -- We have to provide source addresses and derivation paths for this transaction.
+            -- It will be used by external party to provide a proof that it has a right to
+            -- spend this money.
+            let srcAddresses         = extractSrcAddressesFrom unsignedTx
+                srcHdAddresses       = NonEmpty.map (srcAddressToHdAddress db) srcAddresses
+                derivationPaths      = buildDerivationPathsFor srcHdAddresses srcAccId
+                allOutputs           = (NonEmpty.toList $ utxOutputs unsignedTx) ++ changeAddresses
+                unsignedTxWithChange = unsignedTx { utxOutputs = NonEmpty.fromList allOutputs }
+                finalTx              = toRegularTx unsignedTxWithChange
+            return (finalTx, NonEmpty.zip srcAddresses derivationPaths)
+  where
+    -- Take addresses that correspond to inputs of transaction.
+    extractSrcAddressesFrom :: UnsignedTx -> NonEmpty Address
+    extractSrcAddressesFrom unsignedTx =
+        NonEmpty.map (\(_, outAux) -> txOutAddress . toaOut $ outAux) $ utxOwnedInputs unsignedTx
+
+    buildDerivationPathsFor :: NonEmpty HD.HdAddress
+                            -> HD.HdAccountId
+                            -> NonEmpty [DerivationIndex]
+    buildDerivationPathsFor srcAddresses srcAccountId =
+        flip NonEmpty.map srcAddresses $ \(HD.HdAddress srcAddressId _) ->
+            let HD.HdAccountId _ (HD.HdAccountIx srcAccIndex)  = srcAccountId
+                HD.HdAddressId _ (HD.HdAddressIx srcAddrIndex) = srcAddressId
+            in [srcAccIndex, srcAddrIndex]
+
+    -- Convert 'UnsignedTx' to 'Tx', this regular 'Tx'
+    -- will be signed later (technically - the hash of the
+    -- transaction will be signed).
+    toRegularTx :: UnsignedTx -> Tx
+    toRegularTx unsignedTx = UnsafeTx inputs outputs attribs
+      where
+        inputs  = NonEmpty.map fst $ utxOwnedInputs unsignedTx
+        outputs = NonEmpty.map toaOut $ utxOutputs unsignedTx
+        attribs = utxAttributes unsignedTx
+
+-- | Converts source address to 'HdAddress'.
+-- Since 'srcAddress' is returned by 'newUnsignedTransaction'
+-- we assume that this address is valid so we definitely have
+-- corresponding 'HdAddress'.
+srcAddressToHdAddress :: DB
+                      -> Address
+                      -> HD.HdAddress
+srcAddressToHdAddress db srcAddress =
+    either (error "Impossible happened: source address doesn't have corresponding HD-address!")
```

Comment:
Well, this is totally possible, if one supplies a ill-formed `UnsignedTransaction`, no? In which case, would you rather prefer having the node blow up or the error being propagated upstream in a proper API error? I vote for the second option 😉 

My recommendation is to get rid of this `srcAddressToHdAddress` function -- it doesn't provide any extra value. Instead, reuse `lookupCardanoAddress` whenever you need to, but wrap this one in an `ExceptT`, like we do for example in the other functions, for example:

```haskell
    hdAddress <- withExceptT SomeDescriptiveError $ exceptT $
                       lookupCardanoAddress db srcAddress
```